index.bs

```diff
@@ -738,17 +738,91 @@ that should be considered in the development of new features, notably:
+Using explicit flags makes it clear
+whether or not the state changes in different error conditions,
+and makes it clear when the state described by the flags is reset.
+
+Whenever these differences could be observable to web content
+and there isn't a good reason to allow variation between implementations,
+it is important that they be specified unambiguously,
+so that implementations interoperate
+and web content doesn't end up
+depending on the behavior of one implementation and failing in another
+(which can cause higher development costs to fix the content for all browsers,
+or costs to users of broken content if it wasn't detected).
+
+However, sequences of steps are not the only way of making the specification precise enough.
+When there are particular characteristics
+that specification authors want to ensure the specification always satisfies,
+it may be better to use a less algorithmic way of formal specification
+that ensures these characteristics are always true.
+Developing these other ways of specifying behavior formally
+is more important when the algorithms would be more repetitive.
+These formal specification techniques can include dependencies on
+languages such as [[WEBIDL]], on grammar productions,
+or the development of new languages for defining pieces of the specification.
+Merely having a formal syntax is not sufficient, though;
+the requirements that the formal definitions place on implementations must still be clearly defined,
+which generally requires specifying some algorithms (or, sometimes, another layer of formal syntax) that define what the formal syntax means,
+and then using the more-readable formal syntax to avoid
+excessive repetition of the less-readable algorithms.
+There are also possibilities in-between, for example,
+if it is desirable to ensure that an algorithm can be implemented with a state machine,
+it may be preferable to specify it in an algorithmic way that incorporates explicit states
+so that the specification authors avoid accidentally creating more states than intended.
+
+When specifying in an algorithmic way,
+it is generally good to match the algorithms that implementations would use,
+even if this is somewhat less convenient for specification authors.
+For example, it is better to specify CSS selector matching
+by moving right to left across <a>combinators</a>,
+since that is what implementations do.
+This avoids the risk of future problems caused by new features
+that would be slightly different (in unimportant ways)
+between implementations based on one algorithm and implementations based on a different one.
+Thus, it avoids small future divergence in behavior between specification and implementations.
+It also helps to keep the costs of specifying features aligned with implementing them,
+since the complexity of adding a feature to the specification is more likely to be related
+to the cost of adding it to implementations.
```

Comment:
This could use a note about not going full JIT in the specification. Basically, there's a tradeoff here around optimizations, where I think you want to be pretty careful (making optimizations a problem for the implementer isn't entirely responsible either, for instance).

Reply:
Maybe the right condition here is that it's good to  follow the algorithms implementations would use, rather than alternatives, when that doesn't significantly increase the complexity of the specification?

Reply:
Yeah, that sounds like the kind of wording to go for. It might be worth referencing https://infra.spec.whatwg.org/#algorithms as well as ideally all specifications with algorithms build on top of that.